### PR TITLE
Make default_data_collator more flexible and deprecate old behavior

### LINKED ISSUE #5049: DataCollator problem
# ❓ Questions & Help

## Details

Hi everybody
I found an error in the following Colab: 
https://colab.research.google.com/drive/1jwXgtOXE8v8_qkiOCbjFQRFC5semK8T7?usp=sharing

Specifically, As far I understand something changed with the implementation of the following snippet:

class T2TDataCollator(DataCollator):
    def collate_batch(self, batch: List) -> Dict[str, torch.Tensor]:
         ..........
<br>

I got the following error: **TypeError: function() argument 1 must be code, not str**

Can you suggest any workarounds?

### PR DESCRIPTION
**Pull request by sgugger**
This PR does two things:

- avoid breaking changes when people had a custom `DataCollator` with a `collate_batch` method (there is still the breaking change with `DataCollator` not being a class anymore)
- makes `default_data_collator` more flexible by handling dicts on top of `InputExamples` classes coming from our examples

This fixes #5049

### CODE REVIEW

File: src/transformers/data/data_collator.py
Comment (sgugger) [inline src/transformers/data/data_collator.py:39]: This things happens four times below, I put `# See comment in the intro` notes to document it but maybe the corresponding code deserves its own function with the code documented there? Let me know what you prefer.

**Review by LysandreJik — APPROVED** [review]
This looks good to me. Great to have integration tests on it as well.

Let's wait for @julien-c to review before merging.

File: src/transformers/data/data_collator.py
Comment (julien-c) [inline src/transformers/data/data_collator.py:43]: This code looks very duplicated but I don't know if we have an easy + performant way of removing the duplication.
Maybe just doing:
```
d = [vars(f) for f in features]
```
at the top and only handle dicts?

Comment (sgugger) [reply src/transformers/data/data_collator.py:43]: Oh smart, should indeed work to remove the dupe code.

Comment (julien-c) [reply src/transformers/data/data_collator.py:43]: I just hope the perf penalty is acceptable, given that this is on the "hot path" of each batch during training. cc @mfuntowicz

Comment (sgugger) [reply src/transformers/data/data_collator.py:43]: Why would it be longer that the current loop over keys of `var(first)` then `getattr`?

**Comment by sgugger** [comment]
Refactored on @julien-c suggestion and removed the little hack to handle tensors and lists of ints (which was cauding a 4x slowdown on my tests). I just use the first features to test if we have a tensor (then stack) or not (then use torch.tensor).

**Comment by julien-c** [comment]
LGTM!

**Review by LysandreJik — APPROVED** [review]
LGTM, nice test

**Pull request merged**
